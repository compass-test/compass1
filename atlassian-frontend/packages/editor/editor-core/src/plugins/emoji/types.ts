import { CreateUIAnalyticsEvent } from '@atlaskit/analytics-next';
import { EmojiProvider, EmojiDescription } from '@atlaskit/emoji';

export interface EmojiPluginOptions {
  createAnalyticsEvent?: CreateUIAnalyticsEvent;
  allowZeroWidthSpaceAfter?: boolean;
  useInlineWrapper?: boolean;
  headless?: boolean;
}

export type EmojiPluginState = {
  emojiProvider?: EmojiProvider;
  asciiMap?: Map<string, EmojiDescription>;
};
