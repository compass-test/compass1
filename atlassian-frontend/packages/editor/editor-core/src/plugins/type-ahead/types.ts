import { Node as PMNode, Fragment } from 'prosemirror-model';
import type { DecorationSet } from 'prosemirror-view';
import type {
  TypeAheadItem,
  TypeAheadItemRenderProps,
} from '@atlaskit/editor-common/provider-factory';
import { TypeAheadAvailableNodes } from '@atlaskit/editor-common/type-ahead';
import { EditorState, Transaction } from 'prosemirror-state';

import type { INPUT_METHOD } from '../analytics/types/enums';
import type { TypeAheadPayload } from '../analytics/types/type-ahead';
import type { CloseSelectionOptions } from './constants';
import type { UiComponentFactoryParams } from '../../types/ui-components';
import type { SelectItemMode } from '@atlaskit/editor-common/type-ahead';
export type { TypeAheadItem, TypeAheadItemRenderProps };

export type TypeAheadInsert = (
  node?: PMNode | Object | string | Fragment,
  opts?: { selectInlineNode?: boolean },
) => Transaction;

export type OnSelectItem = (props: {
  index: number;
  item: TypeAheadItem;
}) => void;

export interface TypeAheadStats {
  startedAt: number;
  endedAt: number;
  keyCount: {
    arrowUp: number;
    arrowDown: number;
  };
}

export interface TypeAheadStatsSerializable extends TypeAheadStats {
  serialize: () => TypeAheadStats;
}

export interface TypeAheadStatsModifier extends TypeAheadStatsSerializable {
  increaseArrowUp: () => void;
  increaseArrowDown: () => void;
}

export interface TypeAheadStatsMobileModifier
  extends TypeAheadStatsSerializable {
  resetTime: () => void;
  closeTime: () => void;
}

export type TypeAheadSelectItem = (
  state: EditorState,
  item: TypeAheadItem,
  insert: TypeAheadInsert,
  meta: {
    mode: SelectItemMode;
    stats: TypeAheadStats;
    query: string;
    sourceListItem: Array<TypeAheadItem>;
  },
) => Transaction | false;

type TypeAheadForceSelectProps = {
  query: string;
  items: Array<TypeAheadItem>;
  editorState: EditorState;
};

export type TypeAheadForceSelect = (
  props: TypeAheadForceSelectProps,
) => TypeAheadItem | undefined;

export type TypeAheadHandler = {
  id: TypeAheadAvailableNodes;
  trigger: string;
  customRegex?: string;
  headless?: boolean;
  forceSelect?: TypeAheadForceSelect;
  onInvokeAnalytics?: TypeAheadPayload;
  onOpen?: (editorState: EditorState) => void;
  getItems: (props: {
    query: string;
    editorState: EditorState;
  }) => Promise<Array<TypeAheadItem>>;
  selectItem: TypeAheadSelectItem;
  dismiss?: (props: {
    editorState: EditorState;
    query: string;
    stats: TypeAheadStats;
  }) => void;
  getHighlight?: (state: EditorState) => JSX.Element | null;
};

export type TypeAheadPluginState = {
  decorationSet: DecorationSet;
  decorationElement: HTMLElement | null;
  typeAheadHandlers: Array<TypeAheadHandler>;
  query: string;
  items: Array<TypeAheadItem>;
  triggerHandler?: TypeAheadHandler;
  selectedIndex: number;
  stats: TypeAheadStatsSerializable | null;
  inputMethod: TypeAheadInputMethod | null;
};

export type OnInsertSelectedItemProps = {
  mode: SelectItemMode;
  index: number;
  query: string;
};

export type OnItemMatchProps = {
  mode: SelectItemMode;
  query: string;
};
export type OnInsertSelectedItem = (props: OnInsertSelectedItemProps) => void;
export type OnItemMatch = (props: OnItemMatchProps) => boolean;

export type OnTextInsertProps = {
  forceFocusOnEditor: boolean;
  setSelectionAt: CloseSelectionOptions;
  text: string;
};
export type OnTextInsert = (props: OnTextInsertProps) => void;

export type TypeAheadInputMethod =
  | INPUT_METHOD.INSERT_MENU
  | INPUT_METHOD.KEYBOARD
  | INPUT_METHOD.QUICK_INSERT
  | INPUT_METHOD.TOOLBAR;

export type InsertionTransactionMeta = (
  editorState: EditorState,
) => Transaction | false;

type PopupMountPoints = Pick<
  UiComponentFactoryParams,
  'popupsMountPoint' | 'popupsBoundariesElement' | 'popupsScrollableElement'
>;
export type PopupMountPointReference = Record<
  'current',
  PopupMountPoints | null
>;

export type CreateTypeAheadDecorations = (
  tr: Transaction,
  options: {
    triggerHandler: TypeAheadHandler;
    inputMethod: TypeAheadInputMethod;
    reopenQuery?: string;
  },
) => {
  decorationSet: DecorationSet;
  decorationElement: HTMLElement | null;
  stats: TypeAheadStatsSerializable | null;
};

export type RemoveTypeAheadDecorations = (
  decorationSet?: DecorationSet,
) => boolean;
