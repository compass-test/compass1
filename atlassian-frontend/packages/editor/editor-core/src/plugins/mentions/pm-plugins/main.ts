import { Plugin, StateField } from 'prosemirror-state';
import {
  SLI_EVENT_TYPE,
  SMART_EVENT_TYPE,
  buildSliPayload,
  MentionProvider,
} from '@atlaskit/mention/resource';
import {
  ContextIdentifierProvider,
  ProviderFactory,
} from '@atlaskit/editor-common';

import { Dispatch, EventDispatcher } from '../../../event-dispatcher';
import { PortalProviderAPI } from '../../../ui/PortalProvider';
import mentionNodeView from '../nodeviews/mention';
import { mentionPluginKey } from './key';

import type { Command } from '../../../types';
import type { MentionPluginOptions, MentionPluginState } from '../types';

const ACTIONS = {
  SET_PROVIDER: 'SET_PROVIDER',
  SET_CONTEXT: 'SET_CONTEXT',
};

const setProvider = (provider: MentionProvider | undefined): Command => (
  state,
  dispatch,
) => {
  if (dispatch) {
    dispatch(
      state.tr.setMeta(mentionPluginKey, {
        action: ACTIONS.SET_PROVIDER,
        params: { provider },
      }),
    );
  }
  return true;
};

export const setContext = (
  context: ContextIdentifierProvider | undefined,
): Command => (state, dispatch) => {
  if (dispatch) {
    dispatch(
      state.tr.setMeta(mentionPluginKey, {
        action: ACTIONS.SET_CONTEXT,
        params: { context },
      }),
    );
  }
  return true;
};

export function createMentionPlugin(
  dispatch: Dispatch,
  providerFactory: ProviderFactory,
  portalProviderAPI: PortalProviderAPI,
  eventDispatcher: EventDispatcher,
  fireEvent: (payload: any) => void,
  options?: MentionPluginOptions,
) {
  let mentionProvider: MentionProvider;

  const sendAnalytics = (
    event: string,
    actionSubject: string,
    action: string,
    attributes?: {
      [key: string]: any;
    },
  ): void => {
    if (event === SLI_EVENT_TYPE || event === SMART_EVENT_TYPE) {
      fireEvent(buildSliPayload(actionSubject, action, attributes));
    }
  };

  return new Plugin({
    key: mentionPluginKey,
    state: {
      init() {
        return {};
      },
      apply(tr, pluginState) {
        const { action, params } = tr.getMeta(mentionPluginKey) || {
          action: null,
          params: null,
        };

        let newPluginState = pluginState;

        switch (action) {
          case ACTIONS.SET_PROVIDER:
            newPluginState = {
              ...pluginState,
              mentionProvider: params.provider,
            };
            dispatch(mentionPluginKey, newPluginState);

            return newPluginState;
          case ACTIONS.SET_CONTEXT:
            newPluginState = {
              ...pluginState,
              contextIdentifierProvider: params.context,
            };
            dispatch(mentionPluginKey, newPluginState);
            return newPluginState;
        }

        return newPluginState;
      },
    } as StateField<MentionPluginState>,
    props: {
      nodeViews: {
        mention: mentionNodeView(
          portalProviderAPI,
          eventDispatcher,
          providerFactory,
          options,
        ),
      },
    },
    view(editorView) {
      const providerHandler = (
        name: string,
        providerPromise?: Promise<MentionProvider | ContextIdentifierProvider>,
      ) => {
        switch (name) {
          case 'mentionProvider':
            if (!providerPromise) {
              return setProvider(undefined)(
                editorView.state,
                editorView.dispatch,
              );
            }

            (providerPromise as Promise<MentionProvider>)
              .then((provider) => {
                if (mentionProvider) {
                  mentionProvider.unsubscribe('mentionPlugin');
                }

                mentionProvider = provider;
                setProvider(provider)(editorView.state, editorView.dispatch);

                provider.subscribe(
                  'mentionPlugin',
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  sendAnalytics,
                );
              })
              .catch(() =>
                setProvider(undefined)(editorView.state, editorView.dispatch),
              );
            break;

          case 'contextIdentifierProvider':
            if (!providerPromise) {
              return setContext(undefined)(
                editorView.state,
                editorView.dispatch,
              );
            }
            (providerPromise as Promise<ContextIdentifierProvider>).then(
              (provider) => {
                setContext(provider)(editorView.state, editorView.dispatch);
              },
            );
            break;
        }
        return;
      };

      providerFactory.subscribe('mentionProvider', providerHandler);
      providerFactory.subscribe('contextIdentifierProvider', providerHandler);

      return {
        destroy() {
          if (providerFactory) {
            providerFactory.unsubscribe('mentionProvider', providerHandler);
            providerFactory.unsubscribe(
              'contextIdentifierProvider',
              providerHandler,
            );
          }
          if (mentionProvider) {
            mentionProvider.unsubscribe('mentionPlugin');
          }
        },
      };
    },
  });
}
