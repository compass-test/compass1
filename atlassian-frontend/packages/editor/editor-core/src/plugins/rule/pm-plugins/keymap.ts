import { keymap } from 'prosemirror-keymap';
import { Plugin } from 'prosemirror-state';

import * as keymaps from '../../../keymaps';
import { INPUT_METHOD } from '../../analytics';
import { insertHorizontalRule } from '../commands';

export function keymapPlugin(): Plugin {
  const list = {};

  keymaps.bindKeymapWithCommand(
    keymaps.insertRule.common!,
    insertHorizontalRule(INPUT_METHOD.SHORTCUT),
    list,
  );

  keymaps.bindKeymapWithCommand(keymaps.escape.common!, () => true, list);

  return keymap(list);
}

export default keymapPlugin;
