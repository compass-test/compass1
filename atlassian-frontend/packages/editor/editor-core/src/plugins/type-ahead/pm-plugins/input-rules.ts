import { Schema } from 'prosemirror-model';
import { Plugin } from 'prosemirror-state';

import { leafNodeReplacementCharacter } from '@atlaskit/prosemirror-input-rules';
import { InputRuleWrapper } from '@atlaskit/prosemirror-input-rules';
import { openTypeAheadAtCursor } from '../transforms/open-typeahead-at-cursor';

import { INPUT_METHOD } from '../../analytics';
import { createRule, createPlugin } from '../../../utils/input-rules';
import { TypeAheadHandler } from '../types';
import type { FeatureFlags } from '../../../types/feature-flags';

export function inputRulePlugin(
  schema: Schema,
  typeAheads: TypeAheadHandler[],
  featureFlags: FeatureFlags,
): Plugin | undefined {
  if (!typeAheads || typeAheads.length === 0) {
    return;
  }
  const rules = typeAheads.reduce<InputRuleWrapper[]>((acc, typeAhead) => {
    const trigger = typeAhead.customRegex || typeAhead.trigger;
    if (!trigger) {
      return acc;
    }

    const regex = new RegExp(
      `(^|[.!?\\s${leafNodeReplacementCharacter}])(${trigger})$`,
    );

    acc.push(
      createRule(regex, (state, match) => {
        return openTypeAheadAtCursor({
          triggerHandler: typeAhead,
          inputMethod: INPUT_METHOD.KEYBOARD,
        })(state.tr);
      }),
    );

    return acc;
  }, []);

  const plugin = createPlugin('type-ahead', rules, {
    allowInsertTextOnDocument: false,
    useUnpredictableInputRule: featureFlags.useUnpredictableInputRule,
  });

  return plugin;
}

export default inputRulePlugin;
