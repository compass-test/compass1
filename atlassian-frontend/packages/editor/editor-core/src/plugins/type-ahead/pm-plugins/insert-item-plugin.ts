import { Plugin } from 'prosemirror-state';
import { ACTIONS } from './actions';
import { isInsertionTransaction } from './utils';

export function createPlugin(): Plugin {
  return new Plugin({
    appendTransaction(transactions, _oldState, newState) {
      const insertItemCallback = isInsertionTransaction(
        transactions,
        ACTIONS.INSERT_ITEM,
      );
      if (insertItemCallback) {
        const tr = insertItemCallback(newState);
        if (tr) {
          return tr;
        }
      }
    },
  });
}
