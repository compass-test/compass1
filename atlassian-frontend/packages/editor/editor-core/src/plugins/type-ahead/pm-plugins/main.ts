import { EditorState, Transaction, Plugin } from 'prosemirror-state';
import { DecorationSet } from 'prosemirror-view';
import { InsertTypeAheadStep } from '@atlaskit/adf-schema/steps';
import { EditorReactContext } from '../../../types/editor-react-context';

import { CreateUIAnalyticsEvent } from '@atlaskit/analytics-next';
import { Dispatch } from '../../../event-dispatcher';
import type {
  PopupMountPointReference,
  TypeAheadHandler,
  TypeAheadPluginState,
} from '../types';
import { TYPE_AHEAD_DECORATION_DATA_ATTRIBUTE } from '../constants';
import { ACTIONS } from './actions';
import { pluginKey } from './key';
import { closest } from '../../../utils/dom';
import { createReducer } from './reducer';
import { factoryDecorations } from './decorations';
import { isInsertionTransaction } from './utils';

const hasTypeAheadStep = (tr: Transaction): InsertTypeAheadStep | null => {
  const step = tr.steps.find((step) => step instanceof InsertTypeAheadStep);

  if (!step) {
    return null;
  }

  return step as InsertTypeAheadStep;
};

type Props = {
  reactDispatch: Dispatch;
  popupMountRef: PopupMountPointReference;
  typeAheadHandlers: Array<TypeAheadHandler>;
  reactContext: () => EditorReactContext;
  createAnalyticsEvent?: CreateUIAnalyticsEvent;
};
export function createPlugin({
  reactDispatch,
  reactContext,
  popupMountRef,
  createAnalyticsEvent,
  typeAheadHandlers,
}: Props): Plugin {
  const intl = reactContext().intl;
  const { createDecorations, removeDecorations } = factoryDecorations({
    intl,
    popupMountRef,
    createAnalyticsEvent: createAnalyticsEvent,
  });
  const reducer = createReducer({
    createDecorations,
    removeDecorations,
    typeAheadHandlers,
    popupMountRef,
  });
  return new Plugin<TypeAheadPluginState>({
    key: pluginKey,

    state: {
      init() {
        return {
          typeAheadHandlers,
          query: '',
          decorationSet: DecorationSet.empty,
          decorationElement: null,
          items: [],
          selectedIndex: 0,
          stats: null,
          inputMethod: null,
        };
      },

      apply(tr, currentPluginState, oldEditorState, state) {
        const customStep = hasTypeAheadStep(tr);

        const nextPluginState = reducer(tr, currentPluginState, customStep);

        if (currentPluginState !== nextPluginState) {
          reactDispatch(pluginKey, nextPluginState);
        }

        return nextPluginState;
      },
    },

    appendTransaction(transactions, _oldState, newState) {
      const insertItemCallback = isInsertionTransaction(
        transactions,
        ACTIONS.INSERT_RAW_QUERY,
      );
      if (insertItemCallback) {
        const tr = insertItemCallback(newState);
        if (tr) {
          return tr;
        }
      }
    },

    view() {
      return {
        update(editorView) {},
      };
    },

    props: {
      decorations: (state: EditorState) => {
        return pluginKey.getState(state)?.decorationSet;
      },

      handleDOMEvents: {
        compositionend: (view, event: any) => {
          return false;
        },

        click: (view, event) => {
          const { target } = event;
          // ProseMirror view listen to any click event inside of it
          // When this event is coming from the typeahead
          // we should tell to ProseMirror to sit down and relax
          // cuz we know what we are doing (I hope)
          if (
            target instanceof HTMLElement &&
            closest(
              target,
              `[data-type-ahead=${TYPE_AHEAD_DECORATION_DATA_ATTRIBUTE}]`,
            )
          ) {
            return true;
          }
          return false;
        },
      },
    },
  });
}
