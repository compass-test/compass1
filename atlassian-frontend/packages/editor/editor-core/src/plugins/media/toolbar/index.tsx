import React from 'react';
import { InjectedIntl } from 'react-intl';
import { EditorState } from 'prosemirror-state';
import { findParentNodeOfType, removeSelectedNode } from 'prosemirror-utils';
import RemoveIcon from '@atlaskit/icon/glyph/editor/remove';
import DownloadIcon from '@atlaskit/icon/glyph/download';
import { mediaFilmstripItemDOMSelector } from '@atlaskit/media-filmstrip';
import commonMessages from '../../../messages';
import { Command } from '../../../types';
import {
  FloatingToolbarConfig,
  FloatingToolbarItem,
} from '../../../plugins/floating-toolbar/types';
import { stateKey } from '../pm-plugins/plugin-key';
import { hoverDecoration } from '../../base/pm-plugins/decoration';
import { renderAnnotationButton } from './annotation';
import { getLinkingToolbar, shouldShowMediaLinkToolbar } from './linking';
import buildLayoutButtons from '../../../ui/MediaAndEmbedsToolbar';
import { MediaLinkingState, getMediaLinkingState } from '../pm-plugins/linking';
import { getPluginState as getMediaAltTextPluginState } from '../pm-plugins/alt-text';
import { altTextButton, getAltTextToolbar } from './alt-text';
import { MediaFloatingToolbarOptions } from '../types';
import { MediaPluginState } from '../pm-plugins/types';
import { showLinkingToolbar } from '../commands/linking';
import { LinkToolbarAppearance } from './linking-toolbar-appearance';
import {
  ACTION,
  ACTION_SUBJECT,
  ACTION_SUBJECT_ID,
  addAnalytics,
  EVENT_TYPE,
} from '../../analytics';
import { messages } from '@atlaskit/media-ui';
import { messages as cardMessages } from '../../card/messages';
import { FilePreviewItem } from './filePreviewItem';
import { downloadMedia } from './utils';

const remove: Command = (state, dispatch) => {
  if (dispatch) {
    dispatch(removeSelectedNode(state.tr));
  }
  return true;
};

type RemoveMediaCard = (
  state: EditorState,
  mediaPluginState: MediaPluginState,
) => boolean;

const removeMediaCard: RemoveMediaCard = (
  state: EditorState,
  mediaPluginState: MediaPluginState,
) => {
  const getPos = () => {
    const mediaOffset = state.selection.$from.parentOffset + 1;
    return mediaOffset;
  };
  mediaPluginState.handleMediaNodeRemoval(undefined, getPos);
  return true;
};

const generateMediaCardFloatingToolbar = (
  state: EditorState,
  intl: InjectedIntl,
  mediaPluginState: MediaPluginState,
) => {
  const { mediaGroup } = state.schema.nodes;
  const items: FloatingToolbarItem<Command>[] = [
    {
      type: 'dropdown',
      title: intl.formatMessage(messages.displayThumbnail),
      options: [
        {
          title: intl.formatMessage(cardMessages.inline),
          selected: false,
          onClick: () => {
            return true;
          },
          testId: 'inline-appearance',
        },
        {
          title: intl.formatMessage(messages.displayThumbnail),
          selected: true,
          onClick: () => {
            return true;
          },
          testId: 'thumbnail-appearance',
        },
      ],
    },
    {
      type: 'separator',
    },
    {
      type: 'custom',
      fallback: [],
      render: () => {
        return (
          <FilePreviewItem
            key="editor.media.card.preview"
            mediaPluginState={mediaPluginState}
          />
        );
      },
    },
    { type: 'separator' },
    {
      id: 'editor.media.card.download',
      type: 'button',
      icon: DownloadIcon,
      onClick: () => {
        downloadMedia(mediaPluginState);
        return true;
      },
      title: intl.formatMessage(messages.download),
    },
    { type: 'separator' },
    {
      id: 'editor.media.delete',
      type: 'button',
      appearance: 'danger',
      icon: RemoveIcon,
      onMouseEnter: hoverDecoration(mediaGroup, true),
      onMouseLeave: hoverDecoration(mediaGroup, false),
      onFocus: hoverDecoration(mediaGroup, true),
      onBlur: hoverDecoration(mediaGroup, false),
      title: intl.formatMessage(commonMessages.remove),
      onClick: (state) => {
        return removeMediaCard(state, mediaPluginState);
      },
      testId: 'media-toolbar-remove-button',
    },
  ];

  return items;
};

const generateMediaSingleFloatingToolbar = (
  state: EditorState,
  intl: InjectedIntl,
  options: MediaFloatingToolbarOptions,
  pluginState: MediaPluginState,
  mediaLinkingState: MediaLinkingState,
) => {
  const { mediaSingle } = state.schema.nodes;
  const {
    allowResizing,
    allowAnnotation,
    allowLinking,
    allowAdvancedToolBarOptions,
    allowResizingInTables,
    allowAltTextOnImages,
  } = options;

  let toolbarButtons: FloatingToolbarItem<Command>[] = [];
  if (allowAdvancedToolBarOptions) {
    toolbarButtons = buildLayoutButtons(
      state,
      intl,
      state.schema.nodes.mediaSingle,
      allowResizing,
      allowResizingInTables,
    );
    if (toolbarButtons.length) {
      if (allowAnnotation) {
        toolbarButtons.push({
          type: 'custom',
          fallback: [],
          render: renderAnnotationButton(pluginState, intl),
        });
      }
    }

    if (toolbarButtons.length) {
      toolbarButtons.push({ type: 'separator' });
    }

    if (allowLinking && shouldShowMediaLinkToolbar(state)) {
      toolbarButtons.push({
        type: 'custom',
        fallback: [],
        render: (editorView, idx) => {
          if (editorView?.state) {
            const editLink = () => {
              if (editorView) {
                const { state, dispatch } = editorView;
                showLinkingToolbar(state, dispatch);
              }
            };

            const openLink = () => {
              if (editorView) {
                const { state, dispatch } = editorView;
                dispatch(
                  addAnalytics(state, state.tr, {
                    eventType: EVENT_TYPE.TRACK,
                    action: ACTION.VISITED,
                    actionSubject: ACTION_SUBJECT.MEDIA,
                    actionSubjectId: ACTION_SUBJECT_ID.LINK,
                  }),
                );
                return true;
              }
            };

            return (
              <LinkToolbarAppearance
                key={idx}
                editorState={editorView.state}
                intl={intl}
                mediaLinkingState={mediaLinkingState}
                onAddLink={editLink}
                onEditLink={editLink}
                onOpenLink={openLink}
              />
            );
          }
          return null;
        },
      });
    }
  }

  if (allowAltTextOnImages) {
    toolbarButtons.push(altTextButton(intl, state), { type: 'separator' });
  }
  const removeButton: FloatingToolbarItem<Command> = {
    id: 'editor.media.delete',
    type: 'button',
    appearance: 'danger',
    icon: RemoveIcon,
    onMouseEnter: hoverDecoration(mediaSingle, true),
    onMouseLeave: hoverDecoration(mediaSingle, false),
    onFocus: hoverDecoration(mediaSingle, true),
    onBlur: hoverDecoration(mediaSingle, false),
    title: intl.formatMessage(commonMessages.remove),
    onClick: remove,
    testId: 'media-toolbar-remove-button',
  };
  const items: Array<FloatingToolbarItem<Command>> = [
    ...toolbarButtons,
    removeButton,
  ];

  return items;
};

export const floatingToolbar = (
  state: EditorState,
  intl: InjectedIntl,
  options: MediaFloatingToolbarOptions = {},
): FloatingToolbarConfig | undefined => {
  const { media, mediaSingle, mediaGroup } = state.schema.nodes;
  const {
    altTextValidator,
    allowLinking,
    allowAltTextOnImages,
    providerFactory,
    allowMediaInline,
  } = options;
  const mediaPluginState: MediaPluginState | undefined = stateKey.getState(
    state,
  );
  const mediaLinkingState: MediaLinkingState = getMediaLinkingState(state);

  if (!mediaPluginState) {
    return;
  }
  const nodeType = allowMediaInline ? [mediaSingle, media] : [mediaSingle];
  const baseToolbar = {
    title: 'Media floating controls',
    nodeType,
    getDomRef: () => mediaPluginState.element,
  };

  if (
    allowLinking &&
    mediaLinkingState &&
    mediaLinkingState.visible &&
    shouldShowMediaLinkToolbar(state)
  ) {
    const linkingToolbar = getLinkingToolbar(
      baseToolbar,
      mediaLinkingState,
      state,
      intl,
      providerFactory,
    );
    if (linkingToolbar) {
      return linkingToolbar;
    }
  }

  if (allowAltTextOnImages) {
    const mediaAltTextPluginState = getMediaAltTextPluginState(state);
    if (mediaAltTextPluginState.isAltTextEditorOpen) {
      return getAltTextToolbar(baseToolbar, {
        altTextValidator,
      });
    }
  }

  let items: FloatingToolbarItem<Command>[] = [];
  const parentNode = findParentNodeOfType(mediaGroup)(state.selection);
  if (allowMediaInline && parentNode && parentNode.node.type === mediaGroup) {
    const mediaOffset = state.selection.$from.parentOffset + 1;
    baseToolbar.getDomRef = () => {
      const selector = mediaFilmstripItemDOMSelector(mediaOffset);
      return mediaPluginState.element?.querySelector(selector) as HTMLElement;
    };
    items = generateMediaCardFloatingToolbar(state, intl, mediaPluginState);
  } else {
    items = generateMediaSingleFloatingToolbar(
      state,
      intl,
      options,
      mediaPluginState,
      mediaLinkingState,
    );
  }

  return {
    ...baseToolbar,
    items,
  };
};
