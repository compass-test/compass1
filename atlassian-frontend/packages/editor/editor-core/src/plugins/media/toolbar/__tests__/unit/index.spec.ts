jest.mock('prosemirror-utils', () => ({
  ...jest.requireActual<any>('prosemirror-utils'),
  findParentNodeOfType: jest
    .fn()
    .mockReturnValue(() => ({ node: { type: 'mediaGroup' } })),
}));
jest.mock('@atlaskit/media-client', () => ({
  ...jest.requireActual<any>('@atlaskit/media-client'),
  getMediaClient: jest.fn().mockReturnValue({
    file: {
      getCurrentState: jest
        .fn()
        .mockReturnValue({ status: 'processed', name: 'some-file-name' }),
      downloadBinary: jest.fn(),
    },
  }),
}));
import { getMediaClient } from '@atlaskit/media-client';
import { ProviderFactory } from '@atlaskit/editor-common';
import { storyContextIdentifierProviderFactory } from '@atlaskit/editor-test-helpers/context-identifier-provider';
import { createEditorFactory } from '@atlaskit/editor-test-helpers/create-editor';
import {
  media,
  mediaGroup,
  DocBuilder,
} from '@atlaskit/editor-test-helpers/doc-builder';
import { MediaAttributes } from '@atlaskit/adf-schema';
import { IntlProvider } from 'react-intl';
import { MediaOptions } from '../../../../../plugins/media/types';
import { stateKey } from '../../../../../plugins/media/pm-plugins/main';
import { floatingToolbar } from '../../index';
import {
  getFreshMediaProvider,
  temporaryFileId,
  testCollectionName,
} from '../../../../../__tests__/unit/plugins/media/_utils';
import {
  getToolbarItems,
  findToolbarBtn,
} from '../../../../../plugins/floating-toolbar/__tests__/_helpers';
import { MediaPluginState } from '../../../../../plugins/media/pm-plugins/types';

describe('floatingToolbar()', () => {
  const intlProvider = new IntlProvider({ locale: 'en' });
  const { intl } = intlProvider.getChildContext();
  const createEditor = createEditorFactory<MediaPluginState>();

  const editor = (doc: DocBuilder, mediaPropsOverride: MediaOptions = {}) => {
    const contextIdentifierProvider = storyContextIdentifierProviderFactory();
    const mediaProvider = getFreshMediaProvider();
    const providerFactory = ProviderFactory.create({
      contextIdentifierProvider,
      mediaProvider,
    });
    const wrapper = createEditor({
      doc,
      editorProps: {
        media: {
          provider: mediaProvider,
          allowMediaSingle: true,
          ...mediaPropsOverride,
        },
      },
      providerFactory,
      pluginKey: stateKey,
    });
    return wrapper;
  };
  const setup = async (allowMediaInline: boolean) => {
    const attrs: MediaAttributes = {
      id: temporaryFileId,
      type: 'file',
      collection: testCollectionName,
      __fileMimeType: 'image/png',
      __contextId: 'DUMMY-OBJECT-ID',
      width: 100,
      height: 100,
    };

    const doc = mediaGroup(media(attrs)());
    const { editorView, pluginState } = editor(doc, {
      allowMediaInline,
      allowLinking: true,
    });
    await pluginState.setMediaProvider(getFreshMediaProvider());

    // manually override type to pass mediaGroup node checking
    editorView.state.schema.nodes.mediaGroup = 'mediaGroup';
    const toolbar = floatingToolbar(editorView.state, intl, {
      allowMediaInline,
    });
    const items = getToolbarItems(toolbar!, editorView);
    const mediaPluginState: MediaPluginState | undefined = stateKey.getState(
      editorView.state,
    );

    return {
      editorView,
      items,
      mediaPluginState,
    };
  };

  it('should return media group items', async () => {
    const { items } = await setup(true);

    expect(items).toMatchSnapshot();
  });

  it('should not return media group items if allowMediaInline is false', async () => {
    const { items } = await setup(false);

    expect(items).toMatchSnapshot();
  });

  it('should download file when clicking on the download button', async () => {
    const { items, editorView } = await setup(true);
    const downloadButton = findToolbarBtn(items, 'Download');
    const mediaClient = getMediaClient({} as any);

    expect(mediaClient.file.downloadBinary).toHaveBeenCalledTimes(0);
    downloadButton.onClick(editorView.state);
    await mediaClient.file.downloadBinary;
    expect(mediaClient.file.downloadBinary).toHaveBeenCalledTimes(1);
    expect(mediaClient.file.downloadBinary).toBeCalledWith(
      undefined,
      'some-file-name',
      '',
    );
  });

  it('should remove card when clicking on the delete button', async () => {
    const { items, editorView, mediaPluginState } = await setup(true);
    const removeButton = findToolbarBtn(items, 'Remove');
    const spy = jest.spyOn(mediaPluginState!, 'handleMediaNodeRemoval');

    expect(spy).toBeCalledTimes(0);
    removeButton.onClick(editorView.state);
    await new Promise((resolve) => setTimeout(resolve, 1000));

    expect(spy).toBeCalledTimes(1);
  });
});
