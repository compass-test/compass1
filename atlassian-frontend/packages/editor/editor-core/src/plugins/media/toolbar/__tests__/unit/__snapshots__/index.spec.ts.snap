// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`floatingToolbar() should not return media group items if allowMediaInline is false 1`] = `
Array [
  Object {
    "appearance": "danger",
    "icon": [Function],
    "id": "editor.media.delete",
    "onBlur": [Function],
    "onClick": [Function],
    "onFocus": [Function],
    "onMouseEnter": [Function],
    "onMouseLeave": [Function],
    "testId": "media-toolbar-remove-button",
    "title": "Remove",
    "type": "button",
  },
]
`;

exports[`floatingToolbar() should return media group items 1`] = `
Array [
  Object {
    "options": Array [
      Object {
        "onClick": [Function],
        "selected": false,
        "testId": "inline-appearance",
        "title": "Display inline",
      },
      Object {
        "onClick": [Function],
        "selected": true,
        "testId": "thumbnail-appearance",
        "title": "Display thumbnail",
      },
    ],
    "title": "Display thumbnail",
    "type": "dropdown",
  },
  Object {
    "type": "separator",
  },
  Object {
    "fallback": Array [],
    "render": [Function],
    "type": "custom",
  },
  Object {
    "type": "separator",
  },
  Object {
    "icon": [Function],
    "id": "editor.media.card.download",
    "onClick": [Function],
    "title": "Download",
    "type": "button",
  },
  Object {
    "type": "separator",
  },
  Object {
    "appearance": "danger",
    "icon": [Function],
    "id": "editor.media.delete",
    "onBlur": [Function],
    "onClick": [Function],
    "onFocus": [Function],
    "onMouseEnter": [Function],
    "onMouseLeave": [Function],
    "testId": "media-toolbar-remove-button",
    "title": "Remove",
    "type": "button",
  },
]
`;
